controller/server_controller.py

```diff
@@ -125,6 +127,101 @@ def create_fn(labels, logger, name, namespace, spec, uid, **_):
     return {"createdResources": children_uids, "fullServerURL": get_urls(spec)[1]}
 
 
+@kopf.timer(
+    config.api_group,
+    config.api_version,
+    config.custom_resource_name,
+    interval=config.SESSION_IDLE_CHECK_INTERVAL_SECONDS,
+)
+def cull_idle_sessions(body, name, namespace, logger, **kwargs):
+    """
+    Check if a session is idle (has zero open connections in proxy and CPU is below
+    threshold). If the session is idle then annotate the jupyterserver resource with
+    the idle duration. If any sessions have been idle for long enough, then cull them.
+    """
+    try:
+        pod_name = body["status"]["mainPod"]["name"]
+    except KeyError:
+        pod_name = None
+    cpu_usage = get_cpu_usage(pod=pod_name, namespace=namespace)
+    js_server_status = get_js_server_status(body)
+    custom_resource_api = get_api(config.api_version, config.custom_resource_name)
+    idle_seconds = int(
```

Comment:
I think it would be a bit nicer to write this into the status of the custom resource into the body["status"] where all the other status information about the child resources is stored too.